autode/opt/optimisers/crfo.py

```diff
@@ -0,0 +1,206 @@
+"""Constrained rational function optimisation
+
+Notation follows:
+[1] J. Baker, J. Comput. Chem., 18, 8 1080
+[2] J. Baker, J. Comput. Chem., 13, 240 Ž1992
+"""
+import numpy as np
+
+from itertools import combinations
+from autode.log import logger
+from autode.values import GradientRMS, Angle
+from autode.opt.coordinates import CartesianCoordinates, DICWithConstraints
+from autode.opt.coordinates.internals import PIC
+from autode.opt.optimisers.rfo import RFOptimiser
+from autode.opt.optimisers.hessian_update import BFGSDampedUpdate, NullUpdate
+from autode.opt.coordinates.primitives import (Distance, BondAngle, DihedralAngle,
+                                               ConstrainedDistance)
+
+
+class CRFOptimiser(RFOptimiser):
+
+    def __init__(self,
+                 init_alpha: float = 0.05,
+                 *args,
+                 **kwargs):
+        """
+        Constrained rational function optimisation
+
+        -----------------------------------------------------------------------
+        Arguments:
+            init_alpha: Maximum step size
+
+        See Also:
+            :py:meth:`RFOOptimiser <RFOOptimiser.__init__>`
+        """
+        super().__init__(*args, **kwargs)
+
+        self.alpha = float(init_alpha)
+        self._hessian_update_types = [BFGSDampedUpdate, NullUpdate]
+
+    def _step(self) -> None:
```

Comment:
Maybe compressed enough as it can be, but function seems long and contains lots of steps (could some be abstracted?)

Reply:
sort of agree but might leave it. it's still <50 lines and has lots of local variables